### PR TITLE
An efficient method of calling C++ functions from numba/Python using clang++/ctypes/rbc

### PR DESCRIPTION
**Pull request by pearu**
## Text styling

- [ ] The blog is written with [plain language](https://www.plainlanguage.gov/guidelines/) (where relevant).
- [ ] If there are headers, they use the proper header tags in order to do so (with only one level-one header).
- [ ] All links describe where they link to (for example, check the [Quansight labs website](https://labs.quansight.org/)).
- [ ] Any kind of styling that the author uses (for example, bold for emphasis) is consistent throughout the blog.

## Non-text contents

- [ ] All content is represented as text (for example, images need alt text and videos need captions or descriptive transcripts).
- [ ] If there are emojis, there are not more than three in a row.
- [ ] Don't use [flashing gifs or videos](https://www.w3.org/TR/UNDERSTANDING-WCAG20/seizure-does-not-violate.html).
- [ ] If it were to be read as plain text, the blog still makes sense and no information is missing.

### CODE REVIEW

File: posts/2021/10/cxx-numba-interoperability.md [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+functions](https://numba.pydata.org/numba-doc/latest/user/jit.html#compiling-python-code-with-jit)
+--- Python functions that are decorated with
+``numpa.jit(nopython=True)``.
+
+While there exists ways to wrap C++ codes to Python, calling these
+wrappers from Numba jittied functions may not be as straightforward
+and efficient as one would hope. The underlying problem to inefficency
+is that the Python/C++ wrappers are designed to be called on Python
+objects. For instance, with keeping in mind the aim of this post, a
+call to a Numba jitted function would involve converting a Python
```
Comment (HaoZeke) [inline posts/2021/10/cxx-numba-interoperability.md:29]:
```suggestion
call to a Numba JITied function would involve converting a Python
```

File: posts/2021/10/cxx-numba-interoperability.md [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+object to a low-level object, say, to some C/C++ equivalent intrinsic
+type or structure, then converting it back to Python object to be
+passed to the Python/C++ wrapper function, that would convert the
```
Comment (HaoZeke) [inline posts/2021/10/cxx-numba-interoperability.md:32]:
```suggestion
passed to the Python/C++ wrapper function. Then this wrapper function would convert the
```

File: posts/2021/10/cxx-numba-interoperability.md [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+Python object (again) to an equivalent C/C++ type object which can be
+finally passed to the underlying C++ library function. The return
+value of the C++ function would be transformed several times as well,
+just in the opposite direction of the functions calling sequence. In
```
Comment (HaoZeke) [inline posts/2021/10/cxx-numba-interoperability.md:36]:
```suggestion
just in the opposite direction of the function's calling sequence. In
```

File: posts/2021/10/cxx-numba-interoperability.md [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+sum, all these object transformations may build up a considerable
```
Comment (HaoZeke) [inline posts/2021/10/cxx-numba-interoperability.md:37]:
```suggestion
totality, all these object transformations may build up a considerable
```

File: posts/2021/10/cxx-numba-interoperability.md [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+overhead of calling otherwise highly efficient C++ library functions
+from Python.
+
+Another difficulty with calling C++ library functions from Numba
+originates from
+[name-mangling](https://en.wikipedia.org/wiki/Name_mangling) that C++
+compilers apply to function names in order to support function/method
+overloadings as well as other relevant C++ language features. In
+principle, one would be able to call such a C++ library function
+directly from Numba jitted function (using
```
Comment (HaoZeke) [inline posts/2021/10/cxx-numba-interoperability.md:47]:
```suggestion
directly from any Numba JITied function (using the 
```

File: posts/2021/10/cxx-numba-interoperability.md [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+[numba.cfunc](https://numba.pydata.org/numba-doc/latest/user/cfunc.html)
+feature) if one would know how the C++ compiler transforms the
+function name internally . However, the name-mangling algorithm is C++
+compiler vendor dependent, and in practice, it would be hard to
+predict the mangled name in a portable manner.
+
+## cxx2py.py tool
+
+In this post, we'll introduce a straightforward and efficient method
+of calling C++ library functions from Numba jitted functions that
```
Comment (HaoZeke) [inline posts/2021/10/cxx-numba-interoperability.md:57]:
```suggestion
of calling C++ library functions from Numba JITied functions that
```

File: posts/2021/10/cxx-numba-interoperability.md [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+circumvents the name-mangling problem. The method is based on
+determining the addresses of C++ library functions at runtime that
```
Comment (HaoZeke) [inline posts/2021/10/cxx-numba-interoperability.md:59]:
```suggestion
determining the addresses of C++ library functions at runtime which
```

File: posts/2021/10/cxx-numba-interoperability.md [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+together with functions signatures can be used to set up an highly
```
Comment (HaoZeke) [inline posts/2021/10/cxx-numba-interoperability.md:60]:
```suggestion
together with functions signatures are then used to set up a highly
```

File: posts/2021/10/cxx-numba-interoperability.md [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+efficient calling sequence.  This method will require creating a small
+C/C++ wrapper library that contains ``export "C"``-attributed
+functions which return the addresses of C++ libray functions and that
+can be easily called from Python using various techniques, here we use
+[ctypes](https://docs.python.org/3/library/ctypes.html).
+
+A Python script
+[cxx2py.py](https://github.com/pearu/pearu.github.io/tree/main/cxx2py/cxx2py.py)
+is provided that auto-generates, from user-supplied C++ header and
```
Comment (HaoZeke) [inline posts/2021/10/cxx-numba-interoperability.md:69]:
```suggestion
is provided that auto-generates, from a user-supplied C++ header and
```

File: posts/2021/10/cxx-numba-interoperability.md [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+source files, the C/C++ wrapper library as well as a Python
+[ctypes](https://docs.python.org/3/library/ctypes.html) wrapper
+module. The Python module contains ``ctypes`` definitions of C++
+library functions that Numba jitted functions are able to call
```
Comment (HaoZeke) [inline posts/2021/10/cxx-numba-interoperability.md:73]:
```suggestion
library functions that Numba JITied functions are able to call
```

File: posts/2021/10/cxx-numba-interoperability.md [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+.. type: text
+-->
+
+
+# An efficient method of calling C++ functions from numba/Python using clang++/ctypes/rbc
+
+## Introduction
+
+The aim of this document is to explore a method of calling C++ library
+functions from [Numba](http://numba.pydata.org/) [jitted
```
[the remark this answers is not in the document]
Comment (HaoZeke) [reply posts/2021/10/cxx-numba-interoperability.md:19]: I think perhaps `JITied --> Just in time compiled` makes sense over `jitted`.

File: posts/2021/10/cxx-numba-interoperability.md [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+functions](https://numba.pydata.org/numba-doc/latest/user/jit.html#compiling-python-code-with-jit)
+--- Python functions that are decorated with
+``numpa.jit(nopython=True)``.
```
Comment (HaoZeke) [inline posts/2021/10/cxx-numba-interoperability.md:22]:
```suggestion
``numpa.jit(nopython=True)``. Throughout this post we refer to such decorated functions as JITied or Just In Time compiled functions.
```

File: posts/2021/10/cxx-numba-interoperability.md [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+[ctypes](https://docs.python.org/3/library/ctypes.html).
+
+A Python script
+[cxx2py.py](https://github.com/pearu/pearu.github.io/tree/main/cxx2py/cxx2py.py)
+is provided that auto-generates, from user-supplied C++ header and
+source files, the C/C++ wrapper library as well as a Python
+[ctypes](https://docs.python.org/3/library/ctypes.html) wrapper
+module. The Python module contains ``ctypes`` definitions of C++
+library functions that Numba jitted functions are able to call
+directly without requiring the expensive object transformations
```
Comment (HaoZeke) [inline posts/2021/10/cxx-numba-interoperability.md:74]:
```suggestion
directly without requiring the expensive and redundant object transformations
```

File: posts/2021/10/cxx-numba-interoperability.md [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+mentioned above. An alternative to using ``ctypes`` in the Python
+wrapper module would Numba [Wrapper Address Protocol -
```
Comment (HaoZeke) [inline posts/2021/10/cxx-numba-interoperability.md:76]:
```suggestion
wrapper module would be to use the Numba [Wrapper Address Protocol -
```

File: posts/2021/10/cxx-numba-interoperability.md [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+WAP](https://numba.pydata.org/numba-doc/latest/reference/types.html#wrapper-address-protocol-wap)
+that usage should be considered if
```
Comment (HaoZeke) [inline posts/2021/10/cxx-numba-interoperability.md:78]:
```suggestion
. Its usage should be considered if the 
```

File: posts/2021/10/cxx-numba-interoperability.md [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+[ctypes](https://docs.python.org/3/library/ctypes.html) "C++
+expressiveness" turns out to be insufficient.
+
+Currently, the supported features in the ``cxx2py.py`` tool include:
+
+- wrapping of C++ library functions with scalar inputs and return
+  values,
+- supported C++ functions may be defined inside C++ namespaces,
```
Comment (HaoZeke) [inline posts/2021/10/cxx-numba-interoperability.md:86]:
```suggestion
- supporting C++ functions which may be defined inside C++ namespaces,
```

File: posts/2021/10/cxx-numba-interoperability.md [hunk 17]
```diff
@@ -0,0 +1,276 @@
...
+- supported C++ functions may be static class member functions.
```
Comment (HaoZeke) [inline posts/2021/10/cxx-numba-interoperability.md:87]:
```suggestion
- supporting C++ functions which may be static class member functions.
```

File: posts/2021/10/cxx-numba-interoperability.md [hunk 18]
```diff
@@ -0,0 +1,276 @@
...
+
+The ``cxx2py.py`` tool can be extended to support other C++ features
+such as
```
Comment (HaoZeke) [inline posts/2021/10/cxx-numba-interoperability.md:90]:
```suggestion
such as:
```

File: posts/2021/10/cxx-numba-interoperability.md [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+
+- creating C++ class/struct instances from Python,
+- passing C++ class/struct instances around,
```
Comment (HaoZeke) [inline posts/2021/10/cxx-numba-interoperability.md:93]:
```suggestion
- passing C++ class/struct instances to-and-fro between languages,
```

File: posts/2021/10/cxx-numba-interoperability.md [hunk 20]
```diff
@@ -0,0 +1,276 @@
...
+- calling the methods of C++ class/struct instances,
+- supporting pointer types as function inputs and return values,
+- etc.
+
+The ``cxx2py.py`` tool uses ``clang++`` to parse C++ header files and
+to build the C/C++ wrapper shared library. It also uses [RBC - Remote
```
Comment (HaoZeke) [inline posts/2021/10/cxx-numba-interoperability.md:99]:
```suggestion
to build the C/C++ wrapper shared library. It also uses the [RBC - Remote
```

File: posts/2021/10/cxx-numba-interoperability.md (hunk 12 above)
Comment (pearu) [reply posts/2021/10/cxx-numba-interoperability.md:22]: With the use of "Numba compiled functions", I think this suggestion becomes unnecessary.

**Comment by pearu** [comment]
> Have you seen the [cppyy project](https://cppyy.readthedocs.io/en/latest/)? I am not sure how it would work with Numba (there is a mention that [it works](https://cppyy.readthedocs.io/en/latest/philosophy.html?highlight=numba#llvm-dependency) but not more than that). It is meant to be faster and cleaner than pybind11 but has a runtime dependency, like Numba, on the llvm runtime.

Thanks, @mattip for the reference. I think cppyy is worth exploring with respect to numba/C++ connectivity.

Btw, I have added a section "A list of Python/C/C++ connectivity tools". Let me know if some important project is missed.

**Comment by mattip** [comment]
> Btw, I have added a section "A list of Python/C/C++ connectivity tools". Let me know if some important project is missed.

Thanks, it looks even too complete. Maybe only include the ones that are C only like the Python C APi, cython, cffi?

**Comment by rgommers** [comment]
> Btw, I cannot see why Cloudflare Pages build fails. Could anyone post the build failure messages?

I don't seem to have access. That's not great - I don't even know if that's supposed to work (we recently switched to CloudFlare from Netlify, but not very smooth sailing yet).

I usually build locally, should be a matter of creating a new conda env with Python 3.8, pip install from the requirements file in the repo root, and running `nikola auto`.

File: posts/2021/10/cxx-numba-interoperability.md [hunk 21]
```diff
@@ -16,88 +16,88 @@
...
+functions from [Numba](http://numba.pydata.org/) [compiled
 functions](https://numba.pydata.org/numba-doc/latest/user/jit.html#compiling-python-code-with-jit)
 --- Python functions that are decorated with
 ``numpa.jit(nopython=True)``.
 
-While there exists ways to wrap C++ codes to Python, calling these
-wrappers from Numba jittied functions may not be as straightforward
-and efficient as one would hope. The underlying problem to inefficency
+While there exist ways to wrap C++ codes to Python, calling these
+wrappers from Numba JITied functions is often not as straightforward
```
Comment (HaoZeke) [inline posts/2021/10/cxx-numba-interoperability.md:25]:
```suggestion
wrappers from Numba compiled functions is often not as straightforward
```

**Comment by pearu** [comment]
@rgommers I have added the TEASER_END comment as well as copied the referenced files under quansight-labs-site for maximum stability. In case we'll continue working with the cxx2py.py idea, I think it should be turned into a separate Python package anyway.

File: posts/2021/10/cxx-numba-interoperability.md [hunk 22]
```diff
@@ -0,0 +1,286 @@
+<!--
+.. title: An efficient method of calling C++ functions from numba using clang++/ctypes/rbc
+.. slug: cxx-numba-interoperability
+.. date: 2021-10-07 10:30:58
```
[the remark this answers is not in the document]
Comment (tonyfast) [reply posts/2021/10/cxx-numba-interoperability.md:4]: @rgommers @pearu i'm suggesting to schedule this to be published tomorrow. are there any major blockers that would hinder that?

**Review by tonyfast — COMMENTED** [review]
i'd like to publish this tomorrow unless there are any major comments,

**Comment by rgommers** [comment]
> tonyfast merged commit 3f51cc7 into main 3 hours ago

I still had a bunch of review comments on the way, let me submit them now - @pearu feel free to ignore the ones you don't think are relevant, now that the PR is merged.

**Pull request merged**